### PR TITLE
fix: `window.open` causing occasional Node.js crashes

### PR DESCRIPTION
**Pull request by codebytere**
#### Description of Change

Closes https://github.com/electron/electron/issues/37404.
Closes https://github.com/electron/electron/issues/36858.

Refs https://github.com/electron/electron/pull/35999

Fixes an issue where `window.open` can interfere with various aspects of Node.js functionality.

Node.js stores `IsolateData` internally on the `Environment` as a member, whereas Blink stores it using [`V8PerIsolateData`](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/platform/bindings/v8_per_isolate_data.h). This creates a situation where, in the renderer process, a previous `IsolateData` could be used for a new `v8::Context`, which then created a schism in the intended connection between the two. This would, for example, be a problem when using the vm module, because [`ContextifyScript::InstanceOf`](https://github.com/nodejs/node/blob/7f9e60aa1ae9f7da7227c333c1b25809d7ea3c4d/src/node_contextify.cc#L933-L937) would try to check this condition from the `IsolateData` which was mismatched. We fix this by adding a new embedder index and associating it with the context, so that we ensure the `IsolateData` Node.js is using is correctly associated with the current `V8::Context`.

#### Checklist

- [x] PR description included and stakeholders cc'd
- [ ] `npm test` passes
- [x] tests are [changed or added](https://github.com/electron/electron/blob/main/docs/development/testing.md)
- [x] [PR release notes](https://github.com/electron/clerk/blob/main/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/main/README.md#examples).

#### Release Notes

Notes: Fixed an issue where `window.open` can interfere with various aspects of Node.js functionality.

### CODE REVIEW

**Comment by zcbenz** [comment]
I think #37404 and #36858 are essentially different bugs and should be fixed separately.

#37404 consists of 2 bugs:
1. Calling `vm.runInContext` after `window.open` would throw an exception from Node;
2. Throwing an exception in a vm context would crash the exception handler in blink.

#36858 is a race condition that, when closing window one node env marks the isolate as disallow_js during cleanup, while another env node still runs code in the same isolate, which triggers an assertion.

**Comment by deepak1556** [comment]
@zcbenz here is what we found while debugging this PR,

Today Electron allows creation of multiple `Node::Environment` per thread, this can happen in the following situations

1) When `nodeIntegration: true` for main BrowserWindow, we create an environment for the main frame and then when a `window.open` call happens another environment is created via `ElectronRendererClient::DidCreateScriptContext` which gets associated to the child window proxy context.

2) When `nodeIntegration: true` for main BrowserWindow and also `nodeIntegrationInSubFrame: true`, we create an environment for the main frame and when a same origin iframe gets created another environment is created in the same thread via `ElectronRendererClient::DidCreateScriptContext` which gets associated to the iframe window context.

Apart from the above two we allow creating one `Node::Environment` for every web worker thread which is expected and the lifetime of the environment is associated with the worker thread in this case. But in the above two situations the lifetime of the environment is associated with the `v8::Context` since multiple context got created on the same main thread.


On the contrary, in Node.js at a given time only a single `Node::Environment` is associated with a given thread. For the main thread, the environment is associated with [NodeMainInstance::CreateEnvironment](https://github.com/nodejs/node/blob/b38bc9fc890398039f3d97b0710e965d000fa801/src/node_main_instance.cc#L115-L146) and for every Node.js worker thread the environment gets associated with [WorkerThreadData](https://github.com/nodejs/node/blob/b38bc9fc890398039f3d97b0710e965d000fa801/src/node_worker.cc#L355-L362). When there are multiple contexts created per thread, in the case of using vm module the created context gets associated with an existing environment on the thread via [Environment::AssignToContext](https://github.com/nodejs/node/blob/b38bc9fc890398039f3d97b0710e965d000fa801/src/node_contextify.cc#L297-L297) instead of creating a new environment.

Given the current implementation in Electron leads to the race situation where destructing one environment can disallow JS that triggers https://github.com/electron/electron/issues/36858  and also in cases where `vm` module gets used the destruction of an environment can cause the incorrect association of `IsolateData`. I am proposing that we mimic the behavior of Node.js and only allow creation of a single environment per thread. I am also unsure what would be the purpose of having multiple environments per thread since that is not expected today in Node.js. Thoughts ?

Tl:dr; I think this PR is intended to address `Calling vm.runInContext after window.open would throw an exception from Node;` and also the core of the change will address https://github.com/electron/electron/issues/36858

**Comment by zcbenz** [comment]
If we share the Node Environment across multiple contexts, will modules only load in the first context?

File: shell/renderer/electron_renderer_client.cc [hunk 1]
```diff
@@ -41,117 +41,122 @@ void ElectronRendererClient::RenderFrameCreated(
...
+  v8::Maybe<bool> initialized = node::InitializeContext(context);
   CHECK(!initialized.IsNothing() && initialized.FromJust());
 
-  node::Environment* env =
-      node_bindings_->CreateEnvironment(renderer_context, nullptr);
+  // If DidCreateScriptContext is called and we've already created a Node.js
+  // Environment, then we're in the same process with a new V8::Context. If
+  // render_frame is the main frame, then we should replace it. Otherwise, we
+  // should assign the existing Environment to the new V8::Context.
+  if (env_ && !render_frame->IsMainFrame()) {
```
Comment (zcbenz) [inline shell/renderer/electron_renderer_client.cc:95]: The frame in child window opened by `window.open` is also a main frame, with this change there are still multiple environments created after calling `window.open`, it is just the new code no longer does cleanup for the old environment.

Another case is when you reload a page, this code still creates a new environment for the new context, and does not clean up the old one.

Comment (zcbenz) [reply shell/renderer/electron_renderer_client.cc:95]: > Can you point where is this handled in chromium ?

I don't have a reference to the code, this reloading behavior caused some very nasty bugs in `remote` module so it burnt into my memory 😕. You can find some tracings of how js contexts get released during reloads in this issue: https://github.com/electron/electron/issues/8205#issuecomment-403377520.

**Comment by zcbenz** [comment]
Another possible problem with sharing Node Env between contexts is, Node's module system is cached per Env. So if a sub-frame uses the same Env with the main frame, calling `require('./app-init')` in the sub-frame will return the cached object from the context in the main frame, instead of loading the `app-init.js` in the sub-frame's context.

I haven't followed recent changes in Node so this might not be true now, but you might want to do some checks to make sure the module system is sane when sharing Env.

**Comment by gpetrov** [comment]
@codebytere any further developments here? We are desparately waiting for this fix to land.

File: shell/common/node_bindings.cc [hunk 2]
```diff
@@ -518,8 +520,13 @@ node::Environment* NodeBindings::CreateEnvironment(
...
 
-  if (!isolate_data_)
-    isolate_data_ = node::CreateIsolateData(isolate, uv_loop_, platform);
+  auto* isolate_data = context->GetAlignedPointerFromEmbedderData(
+      ELECTRON_CONTEXT_EMBEDDER_DATA_INDEX);
+  if (!isolate_data) {
+    isolate_data = node::CreateIsolateData(isolate, uv_loop_, platform);
+    context->SetAlignedPointerInEmbedderData(
+        ELECTRON_CONTEXT_EMBEDDER_DATA_INDEX, static_cast<void*>(isolate_data));
+  }
```
Comment (deepak1556) [inline shell/common/node_bindings.cc:529]: If we are creating a new environment shouldn't this always create a new isolate data ?

```suggestion
  auto* isolate_data = node::CreateIsolateData(isolate, uv_loop_, platform);
  context->SetAlignedPointerInEmbedderData(
        ELECTRON_CONTEXT_EMBEDDER_DATA_INDEX, static_cast<void*>(isolate_data));
```

**Review by deepak1556 — COMMENTED** [review]
I might not be able to work on https://github.com/electron/electron/pull/38754#issuecomment-1606685990 anytime soon, unblocking the current solution from @codebytere

File: shell/common/node_bindings.cc (hunk 2 above)
Comment (codebytere) [reply shell/common/node_bindings.cc]: i don't think so, because we're not creating a new environment every time here (or changing that logic) we're just ensuring that contexts are correctly associated with isolate data. If there's already a context with the correct isolate data, we don't need to change that

Comment (deepak1556) [reply shell/common/node_bindings.cc]: The code block is currently inside `NodeBindings::CreateEnvironment` which creates an environment every time it is called for a given context, so the newly created context wouldn't have a `node::IsolateData`.

The current check 

```
auto* isolate_data = context->GetAlignedPointerFromEmbedderData(
      ELECTRON_CONTEXT_EMBEDDER_DATA_INDEX);
  if (!isolate_data) {
}
```

will not hold for reasons mentioned in https://github.com/electron/electron/blob/f959fb0c963701c40f309eb8256530f788d5851e/patches/chromium/fix_harden_blink_scriptstate_maybefrom.patch#L26-L40, the failing 32-bit tests confirms this as well. So creating a new isolate data would be an easier route to avoid this and also matches the concept of creating an environment per context. Since in an ideal world both `node::Environment` and `node::IsolateData` are per isolate data structures.

Comment (codebytere) [reply shell/common/node_bindings.cc]: addressed!

File: shell/renderer/electron_renderer_client.cc [hunk 3]
```diff
@@ -144,8 +145,8 @@ void ElectronRendererClient::WillReleaseScriptContext(
 
   node::FreeEnvironment(env);
   if (node_bindings_->uv_env() == nullptr) {
```
Comment (deepak1556) [inline shell/renderer/electron_renderer_client.cc:147]: Given now we create isolate data for every environment, we should destruct the corresponding isolate data to avoid leaks. `node_bindings_->uv_env() == nullptr` check is not needed.

```
 node::FreeIsolateData(node_bindings_->isolate_data(context));
 node_bindings_->clear_isolate_data(context);
```

Comment (deepak1556) [reply shell/renderer/electron_renderer_client.cc:147]: Isn't that because you are accessing the environment after it has been destroyed,

```
node::FreeEnvironment(env);
  if (node_bindings_->uv_env() == nullptr) {
    node::FreeIsolateData(node_bindings_->isolate_data(env->context()));
    node_bindings_->clear_isolate_data(env->context());
  }
```

should instead be

```
v8::Local<v8::Context> context = env->context();
node::FreeEnvironment(env);
node::FreeIsolateData(node_bindings_->isolate_data(context));
node_bindings_->clear_isolate_data(context);
```

Comment (codebytere) [reply shell/renderer/electron_renderer_client.cc:147]: i'm clearly not firing on all cylinders this morning 🙃  corrected.

Comment (codebytere) [reply shell/renderer/electron_renderer_client.cc:147]: I initially thought so too, yes, but if i do that i see the following:

<details><summary>Details</summary>
<p>

```
#
# Fatal error in ../../v8/src/objects/tagged-impl.h, line 142
# Debug check failed: kCanBeWeak || (!IsSmi() == HAS_STRONG_HEAP_OBJECT_TAG(ptr_)).
#
#
#
#FailureMessage Object: 0x16dae0b580   Electron Framework                  0x000000011c4c8a30 base::debug::CollectStackTrace(void**, unsigned long) + 28
1   Electron Framework                  0x000000011c4b9824 base::debug::StackTrace::StackTrace() + 24
2   Electron Framework                  0x000000011dda2978 gin::(anonymous namespace)::PrintStackTrace() + 40
3   Electron Framework                  0x000000011db63bb4 V8_Fatal(char const*, int, char const*, ...) + 268
4   Electron Framework                  0x000000011db63858 std::__Cr::enable_if<!std::is_function<std::__Cr::remove_pointer<unsigned char>::type>::value && !std::is_enum<unsigned char>::value && has_output_operator<unsigned char, v8::base::CheckMessageStream>::value, std::__Cr::basic_string<char, std::__Cr::char_traits<char>, std::__Cr::allocator<char>>>::type v8::base::PrintCheckOperand<unsigned char>(unsigned char) + 0
5   Electron Framework                  0x00000001191d3a98 v8::Utils::OpenHandle(v8::Context const*, bool) + 156
6   Electron Framework                  0x00000001191d6d14 v8::Context::GetNumberOfEmbedderDataFields() + 28
7   Electron Framework                  0x0000000117d8fff0 electron::NodeBindings::isolate_data(v8::Local<v8::Context>) const + 28
8   Electron Framework                  0x0000000117d95960 electron::WebWorkerObserver::~WebWorkerObserver() + 136
9   Electron Framework                  0x0000000117d95808 electron::WebWorkerObserver::Create() + 104
10  Electron Framework                  0x0000000117d90124 electron::ElectronRendererClient::WorkerScriptReadyForEvaluationOnWorkerThread(v8::Local<v8::Context>) + 88
11  Electron Framework                  0x000000011e02a17c blink::WorkerOrWorkletScriptController::PrepareForEvaluation() + 364
12  Electron Framework                  0x000000011e0296d4 blink::WorkerOrWorkletScriptController::Initialize(blink::KURL const&) + 1156
13  Electron Framework                  0x00000001203b61bc blink::PaintWorkletGlobalScope::Create(blink::LocalFrame*, std::__Cr::unique_ptr<blink::GlobalScopeCreationParams, std::__Cr::default_delete<blink::GlobalScopeCreationParams>>, blink::WorkerReportingProxy&) + 104
14  Electron Framework                  0x00000001203b8b34 blink::PaintWorkletGlobalScopeProxy::PaintWorkletGlobalScopeProxy(blink::LocalFrame*, blink::WorkletModuleResponsesMap*, unsigned long) + 1116
15  Electron Framework                  0x00000001203b499c blink::PaintWorklet::CreateGlobalScope() + 292
16  Electron Framework                  0x000000011effcc80 blink::Worklet::FetchAndInvokeScript(blink::KURL const&, WTF::String const&, blink::WorkletPendingTasks*) + 344
17  Electron Framework                  0x000000011effe094 base::internal::Invoker<base::internal::BindState<void (blink::Worklet::*)(blink::KURL const&, WTF::String const&, blink::WorkletPendingTasks*), cppgc::internal::BasicPersistent<blink::Worklet, cppgc::internal::StrongPersistentPolicy, cppgc::internal::IgnoreLocationPolicy, cppgc::internal::DisabledCheckingPolicy>, blink::KURL, blink::V8RequestCredentials, cppgc::internal::BasicPersistent<blink::WorkletPendingTasks, cppgc::internal::StrongPersistentPolicy, cppgc::internal::IgnoreLocationPolicy, cppgc::internal::DisabledCheckingPolicy>>, void ()>::RunOnce(base::internal::BindStateBase*) + 116
18  Electron Framework                  0x000000011aee234c WTF::ThreadCheckingCallbackWrapper<base::OnceCallback<void ()>, void ()>::Run() + 80
19  Electron Framework                  0x000000011c44f704 base::TaskAnnotator::RunTaskImpl(base::PendingTask&) + 308
20  Electron Framework                  0x000000011c476e64 base::sequence_manager::internal::ThreadControllerWithMessagePumpImpl::DoWorkImpl(base::LazyNow*) + 1164
21  Electron Framework                  0x000000011c476494 base::sequence_manager::internal::ThreadControllerWithMessagePumpImpl::DoWork() + 104
22  Electron Framework                  0x000000011c40aa28 base::MessagePumpDefault::Run(base::MessagePump::Delegate*) + 152
23  Electron Framework                  0x000000011c477cb8 base::sequence_manager::internal::ThreadControllerWithMessagePumpImpl::Run(bool, base::TimeDelta) + 472
24  Electron Framework                  0x000000011c430d90 base::RunLoop::Run(base::Location const&) + 396
25  Electron Framework                  0x0000000120ac04c4 content::RendererMain(content::MainFunctionParams) + 1408
26  Electron Framework                  0x0000000117f112d0 content::RunOtherNamedProcessTypeMain(std::__Cr::basic_string<char, std::__Cr::char_traits<char>, std::__Cr::allocator<char>> const&, content::MainFunctionParams, content::ContentMainDelegate*) + 696
27  Electron Framework                  0x0000000117f12030 content::ContentMainRunnerImpl::Run() + 724
28  Electron Framework                  0x0000000117f10584 content::RunContentProcess(content::ContentMainParams, content::ContentMainRunner*) + 1228
29  Electron Framework                  0x0000000117f107e8 content::ContentMain(content::ContentMainParams) + 112
30  Electron Framework                  0x0000000117bbed9c ElectronMain + 128
31  Electron Helper (Renderer)          0x000000010231ca00 main + 224
32  dyld                                0x000000019598bf28 start + 2236
Renderer process crashed - see https://www.electronjs.org/docs/tutorial/application-debugging for potential debugging information.
```

</p>
</details> 

Which i'm looking at right now.

Comment (deepak1556) [reply shell/renderer/electron_renderer_client.cc:147]: Try moving the contents of `~WebWorkerObserver` to `WebWorkerObserver::ContextWillDestroy` to avoid using context after it is signaled for destruction.

Comment (codebytere) [reply shell/renderer/electron_renderer_client.cc:147]: I also tried that in an earlier iteration of this PR - that causes a uv handle closing crash.

Comment (deepak1556) [reply shell/renderer/electron_renderer_client.cc]: Addressed in https://github.com/electron/electron/pull/38754/commits/0337db2408b5dde7fae5619bed3b91cbf47fbfce

**Review by zcbenz — APPROVED** [review]

**Comment by gpetrov** [comment]
@codebytere what is the status here? Seems not really merged with the main - are there still any issues? Like with the Appveyor?

**Comment by KochiyaOcean** [comment]
Is there any plan to backport this fix to the previous version? 26.x is still not stable and the version below 25 still crashes

**Comment by gpetrov** [comment]
Well the fix is still not merged fully, we all hope that @codebytere manage to merge it soon to all supported versions as it is desperately needed

**Comment by gpetrov** [comment]
@codebytere any news on this merge as it seems it isn't even included in the final electron 26.

We really desperately need it and can't move on electron 23+ because of this issue...

**Comment by gpetrov** [comment]
@zcbenz could you please assist in merging this?

**Comment by gpetrov** [comment]
@codebytere could you please continue working on this issue as it isn't merged yet. It is essential to get this fixed so we can move from Electron 22 to newer versions.

**Comment by erikjalevik** [comment]
Thanks to @deepak1556 and the other participants in this thread for shedding light on some obscure issues that have been plaguing my app since the beginning. Since the app is strictly desktop only, and quite a heavy user of Node fs APIs, I have `nodeIntegration` enabled in renderer windows.

Pre-Electron 13, the issues mostly surfaced as strange Node failures and hangs after a window reload, especially around child windows and iframes. But they could mostly be worked around by fiddling with the (now deprecated) `app.allowRendererProcessReuse` flag. 

After finally managing to refactor away `remote` usage, I managed to upgrade through to version 29, and initially everything seemed to work fine, including window reloads and opening child windows. However, after some more extensive testing, I've sadly concluded that the frequency of renderer crashes seen just after closing a child window opened with `window.open`, is currently a blocker for releasing a new version. Based on the stack traces and console output, it seems very likely that this is caused by the same underlying issue as https://github.com/electron/electron/issues/36858.

Therefore, I would like to ask if you can think of any other workarounds except avoiding use of `window.open` altogether? Since it's a race condition, would it maybe be possible to reliably work around it for now by somehow timing any Node interactions differently around the point of closing child windows?
